docs/proposal-new-output-syntax.md

```diff
@@ -0,0 +1,769 @@
+
+## Path Branching
+
+### Concurrent Paths 
+
+Concurrent paths can trigger off of required outputs:
+```
+a:x => b1
+a:y => b2
+b1 & b2 => c  # join
+```
+
+### Alternate Paths
+
+Alternate paths must branch off of optional outputs, and must be joined with
+OR triggers because only one side or the other will run:
+
+```
+a:x? => b1
+a:y? => b2
+b1 | b2 => c  # join
+```
+Note joining can trigger off of required outputs, because the output is
+only "required" if the task actually runs.
+
+**Failure recovery** is a common example of this:
+```
+a? => b1  # success path
+a:fail? => b2  # failure path
+b1 | b2 => c  # join
+```
+
+Note that an artificial dependency can be used to *ensure that at least one
+of the alternate paths gets taken*:
+```
+a:x? => b1
+a:y? => b2
+b1 | b2 => c  # join
+a => c # ARTIFICIAL DEPENDENCY
+```
+This forces `c` to be spawned as an unsatisfied prerequisite waiting on `b1` or
+`b2` even if neither branch runs. If the alternate paths don't actually need to
+join, `c` can be a dummy task.
+
+*See Appendix Output Completion Expressions for a possible follow-up that
+would allow us to flag `a` as incomplete here and avoid the artificial
+dependency.*
+
+## Failed Tasks
+
+No special handling of failed tasks is required.
+- if success is required, a failed task automatically gets retained in the task
+  pool as incomplete.
+- if success is optional (which would usually imply that optional failure is
+  explicitly handled), a failed task can be removed as complete
+
+No special syntax (e.g. `c:fail => NULL`) or suicide triggers (e.g. `c:fail =>
+!c`) are needed for leaf tasks either, if we don't care if they succeed or fail:
+```
+a => b => c?  # success of c is optional
+```
+
+## Family Triggers
+
+The new optional output syntax cannot be used with family collective psuedo
+output triggers because it is difficult to interpret that in terms of the
+success status of each individual member, which is what we need.
+
+Instead we will set different optional/required defaults for each collective
+pseudo output:
+
+```
+A:succeed-all => b  # member success required
+A:succeed-any => b  # member success/failure optional
+
+A:fail-all => b  # member failure required 
+A:fail-any => b  # member success/failure optional
+
+A:finish-all => b  # member success optional
+A:finish-any => b  # member success optional
+
+A:start-all => b  # N/A
+A:start-any => b  # N/A
+
+A:submit-all => b  # member submit success required
+A:submit-any => b  # member success optional
+
+A:submit-fail-all => b  # member submit-fail required
+A:submit-fail-any => b  # member submit/submit-fail optional
+
+A:x-all => b  # member x required
+A:x-any => b  # member x optional
+```
+
+To mess with the defaults you can refer to member outputs collectively or
+individually, anywhere in the graph, e.g.:
+```
+# Family outputs (note these cannot be family triggers)
+A?  # member success optional
+A:x?  # member x optional
```

Comment:
Is it worth supporting this special syntax when you can use , e.g, `A:succeed-any` for the same purpose?
Note that you'd have to use `A:fail-all` to set the default to failure required.

Also, we need to define what happens when 2 family triggers imply different defaults:
```
A => b
A:fail-all => c
```
In the case of mixed defaults I think we should interpret that to mean the default is optional in all cases.

Reply:
We don't actually support `A => b` BTW, have to be explicit with family triggers.

But I agree on "mixed defaults" - forgot to add that to the proposal sorry... for next update

Reply:
What about the special syntax? (no big deal)

Reply:
I was thinking this "special syntax" would be more straightforward and intuitive than using the family pseudo-output trigger syntax just to set output type via the defaults (which vary according to family trigger).  However, not entirely sure about that so I've done the implementation without the special syntax, at least initially.

Reply:
text updated